packages/next/src/client/components/router-reducer/ppr-navigations.ts

```diff
@@ -466,55 +488,93 @@ function createCacheNodeOnNavigation(
-  // the server.
+  // Even we're rendering inside the "new" part of the target tree, we may have
+  // a locally cached segment that we can reuse. This may come from either 1)
+  // the CacheNode tree, which lives in React state and is populated by previous
+  // navigations; or 2) the prefetch cache, which is a separate cache that is
+  // populated by prefetches.
+  let rsc: React.ReactNode
+  let loading: LoadingModuleData | Promise<LoadingModuleData>
+  let head: HeadData | null
+  let cacheNodeNavigatedAt: number
   if (
-    // Check if the segment data is partial
-    isPrefetchRscPartial ||
-    // Check if the head is partial (only relevant if this is a leaf segment)
-    (isPrefetchHeadPartial && isLeafSegment)
+    existingCacheNode !== undefined &&
+    // DYNAMIC_STALETIME_MS defaults to 0, but it can be increased using
+    // the experimental.staleTimes.dynamic config. When set, we'll avoid
+    // refetching dynamic data if it was fetched within the given threshold.
+    existingCacheNode.navigatedAt + DYNAMIC_STALETIME_MS > navigatedAt
   ) {
-    // We only have partial data from this segment. Like missing segments, we
-    // must request the full data from the server.
+    // We have an existing CacheNode for this segment, and it's not stale. We
+    // should reuse it rather than request a new one.
+    rsc = existingCacheNode.rsc
+    loading = existingCacheNode.loading
+    head = existingCacheNode.head
+
+    // Don't update the navigatedAt timestamp, since we're reusing stale data.
+    cacheNodeNavigatedAt = existingCacheNode.navigatedAt
+  } else if (prefetchData !== null) {
+    // There's no existing CacheNode for this segment, but we do have prefetch
+    // data. If the prefetch data is fully static (i.e. does not contain any
+    // dynamic holes), we don't need to request it from the server.
+    rsc = prefetchData[1]
+    loading = prefetchData[3]
+    head = isLeafSegment ? possiblyPartialPrefetchHead : null
+    // Even though we're accessing the data from the prefetch cache, this is
+    // conceptually a new segment, not a reused one. So we should update the
+    // navigatedAt timestamp.
+    cacheNodeNavigatedAt = navigatedAt
+    const isPrefetchRscPartial = prefetchData[4]
+    if (
+      // Check if the segment data is partial
+      isPrefetchRscPartial ||
+      // Check if the head is partial (only relevant if this is a leaf segment)
+      (isPrefetchHeadPartial && isLeafSegment)
+    ) {
+      // We only have partial data from this segment. Like missing segments, we
+      // must request the full data from the server.
+      return spawnPendingTask(
+        navigatedAt,
+        routerState,
+        prefetchData,
+        possiblyPartialPrefetchHead,
+        isPrefetchHeadPartial,
+        segmentPath,
+        scrollableSegmentsResult
+      )
+    } else {
+      // The prefetch data is fully static, so we can omit it from the
+      // navigation request.
```

Comment:
The empty block here is a bit misleading, especially as the fall through does in fact return a new cache node creation.

Reply:
What’s misleading about it? It’s just an explanation of what happens this logical branch represents.

Reply:
In my mind, a block should always contain code. An empty else block as far as I understand gives the impression that some code is missing. I'd rather see the comment exist after the if block instead.